### PR TITLE
Client role

### PR DESCRIPTION
**Pull request by cznewt**
Enforce grafana data sources and dashboards from model.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -5,31 +5,179 @@ Grafana
...
+        server:
+          host: grafana.host
+          port: 3000
+          token: token
+
+
+Usage
+=====
+
+There's a difference between JSON dashboard representation and models we us. Lists are replaced by dictionaries to support mergings and interpolations.
```
Comment (simonpasquier) [inline README.rst]: > There's a difference between JSON dashboard representation and models we us.

The end of the sentence doesn't make sense

File: README.rst [hunk 2]
```diff
@@ -5,31 +5,179 @@ Grafana
...
-          metrics2:
-            engine: elasticsearch
-            host: metrics2.domain.com
+          elasticsearch:
+            type: elasticsearch
+            host: log01.domain.com
             port: 80
-            user: test
+            user: admin
+            password: password
```
Comment (simonpasquier) [inline README.rst]: nit but unless you use Shield or some kind of reverse proxy, Elasticsearch doesn't require authentication.

File: grafana/files/grafana.ini [hunk 3]
```diff
@@ -55,29 +55,33 @@ http_port = {{ server.bind.port }}
...
 user = {{ server.database.user }}
 password = {{ server.database.password }}
+{%- endif %}
 
 # For "postgres" only, either "disable", "require" or "verify-full"
 ;ssl_mode = disable
 
 # For "sqlite3" only, path relative to data_path setting
-;path = grafana.db
+{%- if server.database.engine in ["postgresql"] %}
```
Comment (simonpasquier) [inline grafana/files/grafana.ini]: s/postgresql/sqlite/

File: grafana/files/grafana.ini [hunk 4]
```diff
@@ -55,29 +55,33 @@ http_port = {{ server.bind.port }}
...
+path = grafana.db
```
Comment (simonpasquier) [inline grafana/files/grafana.ini]: an absolute path would be good

File: README.rst (hunk 1 above)
Comment (cznewt) [inline README.rst]: you cannot merge/override lists easily

Comment (cznewt) [inline README.rst]: fixed

File: grafana/files/grafana.ini (hunk 3 above)
Comment (cznewt) [inline grafana/files/grafana.ini]: fixed

File: grafana/files/grafana.ini (hunk 4 above)
Comment (cznewt) [inline grafana/files/grafana.ini]: yes, but as sqlite is not recommended at all, i'll pass this

File: grafana/client.sls [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  - user: root
+  - group: root
+
+{%- for datasource_name, datasource in client.datasource.iteritems() %}
+
+grafana_client_datasource_{{ datasource_name }}:
+  grafana_datasource.present:
+  - name: {{ datasource_name }}
+  - type: {{ datasource.type }}
+  - url: http://{{ datasource.host }}:{{ datasource.get('port', 80) }}
```
Comment (fpytloun) [inline grafana/client.sls]: What's this hardcoded http?

File: grafana/files/grafana.ini [hunk 6]
```diff
@@ -55,29 +55,33 @@ http_port = {{ server.bind.port }}
...
 password = {{ server.database.password }}
+{%- endif %}
 
 # For "postgres" only, either "disable", "require" or "verify-full"
 ;ssl_mode = disable
 
 # For "sqlite3" only, path relative to data_path setting
-;path = grafana.db
+{%- if server.database.engine in ["sqlite"] %}
+path = grafana.db
```
Comment (fpytloun) [inline grafana/files/grafana.ini]: Shouldn't be hard-coded location I guess.

File: grafana/files/_grafana.conf [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{%- from "grafana/map.jinja" import client with context %}
+
+grafana_version: 2
+
+grafana:
+  grafana_timeout: 3
+  grafana_token: {{ client.server.token }}
+  grafana_url: 'http://{{ client.server.host }}:{{ client.server.get('port', 80) }}'
```
Comment (fpytloun) [inline grafana/files/_grafana.conf]: Hard-coded http

File: grafana/files/_grafana.conf [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{%- from "grafana/map.jinja" import client with context %}
+
+grafana_version: 2
+
+grafana:
+  grafana_timeout: 3
```
Comment (fpytloun) [inline grafana/files/_grafana.conf]: It would be good to have timeout here and maybe version about configurable

File: grafana/files/_grafana.conf [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+  grafana_token: {{ client.server.token }}
```
Comment (fpytloun) [inline grafana/files/_grafana.conf]: It's not described what's the token about and how it's going to be set up, etc.

File: grafana/map.jinja [hunk 10]
```diff
@@ -7,6 +7,29 @@ Debian:
...
+  session:
+    engine: file
+  auth:
+    engine: application
+  admin:
+    user: admin
+    password: admin
+  allow_sign_up: False
+  allow_org_create: False
+  auto_assign_role: Viewer
```
Comment (simonpasquier) [inline grafana/map.jinja]: Out of curiosity, why did you add the default values to this file instead of metadata/service/server.single.yml?
I see that depending on the formulas, default values are added to the service metadata, map.jinja or both...

File: grafana/files/_grafana.conf (hunk 7 above)
Comment (cznewt) [inline grafana/files/_grafana.conf]: ou yeah :D can you at least once close your eyes a little?

File: grafana/client.sls (hunk 5 above)
Comment (cznewt) [inline grafana/client.sls]: I'll merge it and propose better solution with new salt resources later 2day with your comments integrated.
